### PR TITLE
west: build folder format configuration option

### PR DESCRIPTION
**Pull request by carlescufi**
Add the possibility of configuring the build folder format in west's configuration system. 
The build.dir-fmt configuration option controls how west will create build folders when not specified, the following parameters are currently accepted:
- `board`: The board name
- `source_dir`: The relative path from the CWD to the source directory

If the CWD is below `source_dir` in the directory hierarchy then `source_dir` is set to an empty string.

This means that if one sets:
```
[build]
dir-fmt = build/{board}/{source_dir}
```
Then when building samples/hello_world from zephyr's root for the `reel_board` the build folder will be:
```
./build/reel_board/samples/hello_world
```
but when building it from inside the `samples/hello_world` folder it will instead be:
```
./build/reel_board
```

### CODE REVIEW

**Comment by carlescufi** [comment]
Note: I am open to an additional formatting option like:
`{source-dir-top}`
which would strip the path except for the top-level folder of the source directory. So if you were to build `samples/hello_world` from zephyr's root the build folder would be:

`build/reel_board/hello_world`

**Comment by carlescufi** [comment]
Note 2: `west flash, debug, attach` and company are not currently able to follow the formatting, so you will need to add `-d path/to/build/folder`. I want to discuss with @mbolivar whether it would be possible to fix that.

**Review by dbkinder — APPROVED** [review]
doc change looks OK, thanks.

**Review by nashif — APPROVED** [review]

File: scripts/west_commands/run_common.py [hunk 1]
```diff
@@ -128,16 +128,14 @@ def _build_dir(args, die_if_none=True):
     if args.build_dir:
         return args.build_dir
 
-    cwd = getcwd()
-    default = path.join(cwd, DEFAULT_BUILD_DIR)
-    if is_zephyr_build(default):
-        return default
-    elif is_zephyr_build(cwd):
-        return cwd
+    dir = find_build_dir(None)
```
Comment (mbolivar) [inline scripts/west_commands/run_common.py]:
```
This actually changes the behavoir slightly, since the current working
directory (cwd) will now be checked before the default build folder
('build'). Since builds within builds are not supported this should not
be an issue.
```

Thanks for calling attention to that. This is actually an issue -- it breaks compatibility with documented behavior:

```
  -d BUILD_DIR, --build-dir BUILD_DIR
                        Build directory to obtain runner information from. If
                        not given, this command tries to use build/ and then
                        the current working directory, in that order.
```

I think the helper in zephyr should be considered buggy and search in the order we use for --build-dir documentation.

File: doc/guides/west/build-flash-debug.rst [hunk 2]
```diff
@@ -86,6 +86,18 @@ To use another build directory, use ``--build-dir`` (or ``-d``)::
 
   west build -b <BOARD> --build-dir path/to/build/directory
 
+To change the default build directory from ``build`` to a user-defined format,
+use the ``dir-fmt`` configuration option::
+
+  west config build.dir-fmt "build/{board}/{source_dir}"
```
Comment (mbolivar) [inline doc/guides/west/build-flash-debug.rst]: Is it just me, or would changing this to `{app}` and making the result be the basename of the source directory be easier for most people to use?

```
$ cd ~/zephyrproject
$ west config build.dir-fmt build/{board}/{app}
$ west build -b reel_board zephyr/samples/hello_world  # makes build/hello_world/reel_board 
$ west build -b qemu_x86 zephyr/samples/hello_world    # makes build/hello_world/qemu_x86
```

Especially if I want faster tab completion on `west flash -d`.

File: doc/guides/west/build-flash-debug.rst [hunk 3]
```diff
@@ -86,6 +86,18 @@ To use another build directory, use ``--build-dir`` (or ``-d``)::
...
+
+With the configuration above, if you build :file:`samples/hello_world` from
+the root of zephyr's repository for the ``reel_board`` the build folder will
+be created at :file:`build/reel_board/samples/hello_world`. If you however
+change your current working directory to :file:`samples/hello_world`, the
+build folder will be created at :file:`build/reel_board`, since the relative
+path to the source directory is just :file:`.`.
```
Comment (mbolivar) [inline doc/guides/west/build-flash-debug.rst]: It also would mean the docs here could be shorter, since there's no difference in the results no matter where you run it.

File: scripts/west_commands/build_helpers.py [hunk 4]
```diff
@@ -25,21 +27,42 @@
...
             build_dir = cwd
         else:
-            build_dir = DEFAULT_BUILD_DIR
+            build_dir = config.get('build', 'dir-fmt', fallback=None)
+            log.dbg('config dir-fmt: {}'.format(build_dir))
+            if build_dir:
+                build_dir = _resolve_build_dir(build_dir, cwd, **kwargs)
+            if not build_dir:
+                build_dir = DEFAULT_BUILD_DIR
+    log.dbg('build dir: {}'.format(build_dir))
```
Comment (mbolivar) [inline scripts/west_commands/build_helpers.py]: can you bump this to VERBOSE_EXTREME?

File: scripts/west_commands/build_helpers.py [hunk 5]
```diff
@@ -25,21 +27,42 @@
...
     if dir:
         build_dir = dir
     else:
         cwd = os.getcwd()
         if is_zephyr_build(cwd):
             build_dir = cwd
         else:
-            build_dir = DEFAULT_BUILD_DIR
+            build_dir = config.get('build', 'dir-fmt', fallback=None)
+            log.dbg('config dir-fmt: {}'.format(build_dir))
```
Comment (mbolivar) [inline scripts/west_commands/build_helpers.py]: VERBOSE_EXTREME please

File: scripts/west_commands/build_helpers.py [hunk 6]
```diff
@@ -25,21 +27,42 @@
...
+    is returned.'''
     if dir:
         build_dir = dir
     else:
         cwd = os.getcwd()
         if is_zephyr_build(cwd):
             build_dir = cwd
         else:
-            build_dir = DEFAULT_BUILD_DIR
+            build_dir = config.get('build', 'dir-fmt', fallback=None)
```
Comment (mbolivar) [inline scripts/west_commands/build_helpers.py]: Shouldn't this be tested first (then build/, then cwd, as discussed elswhere)?

File: scripts/west_commands/build_helpers.py [hunk 7]
```diff
@@ -25,21 +27,42 @@
...
+    if source_dir:
+        if escapes_directory(cwd, source_dir):
+            kwargs['source_dir'] = os.path.relpath(source_dir, cwd)
+        else:
+            # no meaningful relative path possible
+            kwargs['source_dir'] = ''
+    try:
+        return fmt.format(**kwargs)
+    except KeyError as k:
+        return None
```
Comment (mbolivar) [inline scripts/west_commands/build_helpers.py]: Slightly more pythonic would be to let the exception bubble up and handle it in find_build_dir

File: scripts/west_commands/build.py [hunk 8]
```diff
@@ -155,7 +155,7 @@ def do_run(self, args, remainder):
                     self.run_cmake = True
         else:
             self.run_cmake = True
-        self._setup_source_dir()
+        self.source_dir = self._find_source_dir()
```
Comment (mbolivar) [inline scripts/west_commands/build.py]: I don't understand the need for this change. The function currently sets self.source_dir to what it finds. But it's being changed to return a directory, which is always being assigned to self.source_dir.

File: scripts/west_commands/build.py [hunk 9]
```diff
@@ -231,7 +234,7 @@ def _setup_build_dir(self):
 
         self.build_dir = build_dir
 
-    def _setup_source_dir(self):
+    def _find_source_dir(self):
         # Initialize source_dir attribute, either from command line argument,
```
Comment (mbolivar) [inline scripts/west_commands/build.py]: This comment is no longer true after this change.

In general looking at this file with fresh eyes, there are places where we're using self.args.source_dir where we should be using self.source_dir (e.g. in _parse_remainder). The former is what the user said, if anything, the latter is what we decided the answer is.

The purpose of this function is to set self.source_dir, but it seems to be being extended to be used differently. The purpose is to find the source directory so it can be used in the source_dir format string when creating the build directory.

This is starting to get a bit messy and hard to follow. I think it would be better combined into a single `_setup_dirs()` function that sets `self.source_dir`, `self.build_dir`, and `self.created_build_dir`. That will clean up the code around the "legacy -s  option" comment as well. This routine should probably also initialize the CMake cache if the build directory exists.

**Review by mbolivar — CHANGES_REQUESTED** [review]
Thanks for this!

File: scripts/west_commands/run_common.py (hunk 1 above)
Comment (carlescufi) [reply scripts/west_commands/run_common.py]: >I think the helper in zephyr should be considered buggy and search in the order we use for --build-dir documentation.

Not sure if I understand you here, you mean `find_build_folder` should be considered buggy and the runners should instead use the 1.`build` 2.`cwd` order?

File: doc/guides/west/build-flash-debug.rst (hunk 2 above)
Comment (carlescufi) [reply doc/guides/west/build-flash-debug.rst]: Sure, no objections. `source_dir` is consistent with other places, but if you think `{app}` is clearer I have no issues changing it

File: doc/guides/west/build-flash-debug.rst (hunk 3 above)
Comment (carlescufi) [reply doc/guides/west/build-flash-debug.rst]: Sure, but I wanted to highlight the fact that the tool works both outside and inside the sample folder

File: scripts/west_commands/build.py (hunk 8 above)
Comment (carlescufi) [reply scripts/west_commands/build.py]: I didn't want to set `self.source_dir` when setting up the build folder. In that case the result is assigned to a local variable, not to a class member.

File: scripts/west_commands/build.py (hunk 9 above)
Comment (carlescufi) [reply scripts/west_commands/build.py]: >In general looking at this file with fresh eyes, there are places where we're using self.args.source_dir where we should be using self.source_dir (e.g. in _parse_remainder). The former is what the user said, if anything, the latter is what we decided the answer is.

I don't think so, in `_parse_remainder()` we're just using `self.args.source_dir` as a temporary variable. That exists because we wanted to keep compatibility with `-s`.
Regarding the cleanup, I agree with you, but I'd rather not do it in this PR. I don't think the changes in this PR worsen the current state of things in any significant way, and I have more changes in mind for the near future that will require the cleanup.

File: scripts/west_commands/build_helpers.py (hunk 6 above)
Comment (carlescufi) [reply scripts/west_commands/build_helpers.py]: Well, that really depends on what we choose to do. I actually used a an old comment of yours as a reference: https://github.com/zephyrproject-rtos/west/issues/124#issuecomment-448064044 but modified it slightly:

>1. -d is not given at the command line (highest precedence)
>2. $PWD is not a zephyr build directory
>3. $PWD/build (the default build directory) is not a zephyr build directory (lowest precedence)
>4. the directory implied by the west.command.build.build-dir-fmt option and command line arguments does not exist (i.e. if west build --board=my-board is given, build-dir-fmt = build/{board}, and build/my-board exists, use that)

File: doc/guides/west/build-flash-debug.rst (hunk 2 above)
Comment (carlescufi) [reply doc/guides/west/build-flash-debug.rst]: Oh sorry, I now understand. See my comment here: https://github.com/zephyrproject-rtos/zephyr/pull/16636#issuecomment-499104299

**Review by erwango — APPROVED** [review]
I'm fine with the proposal:
*with `{source_dir}` being ignored if CWD is below `source_dir`
*using `{source_dir}` vs `{app}` (to avoid potential app name conflicts)

File: scripts/west_commands/run_common.py (hunk 1 above)
Comment (mbolivar) [reply scripts/west_commands/run_common.py]: Yep, as discussed on slack.

File: doc/guides/west/build-flash-debug.rst (hunk 3 above)
Comment (mbolivar) [reply doc/guides/west/build-flash-debug.rst]: What I mean is, using {app} leads to uniform results, so there is no need to explain here "if you're here, then it's this, if you're somewhere else, then it's that".

I'm fine with having both.

File: doc/guides/west/build-flash-debug.rst (hunk 2 above)
Comment (mbolivar) [reply doc/guides/west/build-flash-debug.rst]: I'm OK with both and would rather have a separate {app} than {source_dir[:1]}.

Note that at least for upstream samples, app names tend to be unique. These are the only dupes I could find:

```
      2 ap
      2 echo
      3 philosophers
      2 remote
      2 sensor
      2 timer_synchronization
```

So unless you're literally building all the samples, {app} is not going to collide.

File: scripts/west_commands/build.py (hunk 8 above)
Comment (mbolivar) [reply scripts/west_commands/build.py]: Crap, I saw that and meant to delete this comment before posting the review; sorry.

File: scripts/west_commands/build.py (hunk 9 above)
Comment (mbolivar) [reply scripts/west_commands/build.py]: > I have more changes in mind for the near future that will require the cleanup.

If that means you're willing do to the cleanup I'm OK with it. The current change has crossed my personal OCD threshold ;).

**Review by lemrey — APPROVED** [review]
This is a nice feature, thanks 👍

**Review by mbolivar — CHANGES_REQUESTED** [review]

File: scripts/west_commands/build_helpers.py [hunk 10]
```diff
@@ -21,22 +21,25 @@
...
+Build directory. If not given, {}/ is used; otherwise if the current directory
+is a Zephyr build directory, it is used.'''.format(DEFAULT_BUILD_DIR)
 
 
 def find_build_dir(dir):
     '''Heuristic for finding a build directory.
 
-    If the given argument is truthy, it is returned. Otherwise, if
+    If the given argument is truthy, it is returned. Otherwise if the
+    DEFAULT_BUILD_DIR is a build directory, it is returned. Next, if
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply scripts/west_commands/build_helpers.py]: I think you are reviewing an old revision

File: doc/guides/west/build-flash-debug.rst [hunk 11]
```diff
@@ -35,17 +35,25 @@ use :ref:`west config <west-config-cmd>` to configure its behavior.
 This command attempts to "do what you mean" when run from a Zephyr application
 source or build directory:
 
-- When you run ``west build`` in an existing build directory, the board, source
+- When you run ``west build`` and a Zephyr build directory named as the default
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply doc/guides/west/build-flash-debug.rst]: > Since this is the introduction to the command, I think it's important to have clear wording that says what it does "in a nutshell" and "out of the box" here. This change here is hard to understand. I think we can balance this with the need to make it clear there are other options a bit better.
> 
> I've pushed a patch with some ideas, along with applying the other suggestions in these comments, etc:
> 
> https://github.com/mbolivar/zephyr/tree/build-folder-config

I have taken your changes as-is except for a minor fix (debug copy-paste mistake). I think they are better than what I proposed. That said, I still think the default should be `source_dir` instead of `app`, but I see no consensus here and so I'd rather go with what you think makes sense in this particular case.

File: scripts/west_commands/run_common.py [hunk 12]
```diff
@@ -128,16 +127,19 @@ def _build_dir(args, die_if_none=True):
...
+
+    if dir and is_zephyr_build(dir):
+        return dir
     elif die_if_none:
-        log.die('--build-dir was not given, and neither {} '
-                'nor {} are zephyr build directories.'.
-                format(default, cwd))
+        msg ='--build-dir was not given, '
```
Comment (mbolivar) [inline scripts/west_commands/run_common.py]: whitespace nit

File: scripts/west_commands/build_helpers.py [hunk 13]
```diff
@@ -13,35 +13,70 @@
...
         build_dir = dir
     else:
         cwd = os.getcwd()
-        if is_zephyr_build(cwd):
+        default = config.get('build', 'dir-fmt', fallback=DEFAULT_BUILD_DIR)
+        try:
+            default = _resolve_build_dir(default, cwd, **kwargs)
+            log.dbg('config dir-fmt: {}'.format(default),
+                    level=log.VERBOSE_EXTREME)
+        except KeyError as k:
```
Comment (mbolivar) [inline scripts/west_commands/build_helpers.py]: `as k` with `k` unused is a flake8 warning; to silence it, use `except KeyError:` instead

File: scripts/west_commands/build_helpers.py [hunk 14]
```diff
@@ -13,35 +13,70 @@
...
-    DEFAULT_BUILD_DIR)
+FIND_BUILD_DIR_DESCRIPTION = '''\
+If not given, the default build directory ({}/ unless the
+build.dir-fmt configuration variable is set) and the current directory are
+checked, in that order. If one is a Zephyr build directory, it is used.
+'''.format(DEFAULT_BUILD_DIR)
 
+def _resolve_build_dir(fmt, cwd, **kwargs):
+    # Remove any None values, we do not want 'None' as a string
+    kwargs = {k:v for k,v in kwargs.items() if v is not None}
```
Comment (mbolivar) [inline scripts/west_commands/build_helpers.py]: `{k: v for k, v in kwargs.items() if v is not None}` will keep flake8 happy

File: scripts/west_commands/run_common.py [hunk 15]
```diff
@@ -128,16 +127,19 @@ def _build_dir(args, die_if_none=True):
...
+        if dir:
+            msg = msg + 'and neither {} nor {} are zephyr build directories.'
+        else:
+            msg = msg + ('{} is not a build directory and the default build '
+                         'directory cannot be determined. Check your '
+                         'build.dir-fmt configuration option')
```
Comment (mbolivar) [inline scripts/west_commands/run_common.py]: no need to make a change, but it would be slightly nicer to print the value here if it is not `None`

File: doc/guides/west/build-flash-debug.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (mbolivar) [reply doc/guides/west/build-flash-debug.rst]: OK, thanks for humoring me and for catching that flash/debug copy/paste error :).

File: scripts/west_commands/run_common.py (hunk 15 above)
Comment (carlescufi) [reply scripts/west_commands/run_common.py]: if what is not `None`? `dir`? I check against that and print it above?

**Comment by carlescufi** [comment]
@mbolivar fixed the flake8 issues, thank you!

File: scripts/west_commands/run_common.py (hunk 15 above)
Comment (mbolivar) [reply scripts/west_commands/run_common.py]: `build.dir-fmt`

**Review by mbolivar — APPROVED** [review]
LGTM, thanks for this feature!

File: scripts/west_commands/run_common.py (hunk 15 above)
Comment (carlescufi) [reply scripts/west_commands/run_common.py]: oh right. That would force me to load it from outside `build_helpers.py`. I'd rather not do that.

Comment (mbolivar) [reply scripts/west_commands/run_common.py]: No worries, then.
